pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java

```diff
@@ -163,48 +181,44 @@ public synchronized void refreshSegment(String segment) {
       if (filterExpression == null) {
         return segments;
       }
-      Set<String> selectedSegments = new HashSet<>();
-      for (String segment : segments) {
-        PartitionInfo partitionInfo = _partitionInfoMap.get(segment);
-        if (partitionInfo == null || partitionInfo == INVALID_PARTITION_INFO || isPartitionMatch(filterExpression,
-            partitionInfo)) {
-          selectedSegments.add(segment);
-        }
-      }
-      return selectedSegments;
+      return pruneSegments((partitionInfo, cachedPartitionFunction) -> isPartitionMatch(filterExpression,
+        partitionInfo, cachedPartitionFunction));
     } else {
       // PQL
       FilterQueryTree filterQueryTree = RequestUtils.generateFilterQueryTree(brokerRequest);
       if (filterQueryTree == null) {
         return segments;
       }
-      Set<String> selectedSegments = new HashSet<>();
-      for (String segment : segments) {
-        PartitionInfo partitionInfo = _partitionInfoMap.get(segment);
-        if (partitionInfo == null || partitionInfo == INVALID_PARTITION_INFO || isPartitionMatch(filterQueryTree,
-            partitionInfo)) {
-          selectedSegments.add(segment);
-        }
+      return pruneSegments((partitionInfo, cachedPartitionFunction) -> isPartitionMatch(filterQueryTree, partitionInfo, cachedPartitionFunction));
+    }
+  }
+
+  private Set<String> pruneSegments(java.util.function.BiFunction<PartitionInfo, CachedPartitionFunction,  Boolean> partitionMatchLambda) {
+    Set<String> selectedSegments = new HashSet<>();
+    CachedPartitionFunction cachedPartitionFunction = new CachedPartitionFunction();
```

Comment:
Why are we creating a cache per query? If we just need to cache the partition id, `Map<PartitionFunction, Map<String, Integer>>` should be enough

Reply:
The Cache/LookupTable here is not for our Segment names. It is for the Value -> PartitionNo in each SQL statement. E.g., if we have something like `WHERE account_key = 'a7cd24e'` and we need to compute `MurmurHash.getPartition('a7cd24e')` repeatedly for each segment. If we have 10k segments, we may end up computing  `MurmurHash.getPartition('a7cd24e')` 10k times. 
Because any value can be passed in SQL statement, it makes sense to create the cache for each SQL

Reply:
curious any benefit to keep cache across queries? were partitioning values across queries kinda repetitive?